### ISSUE TITLE
[Feature]: Store Trakt.tv authentication details outside of config.yml

### ISSUE THREAD

**Issue by rfgamaral** [Kometa-Team/Kometa#1019]
### Is your feature request related to a problem? Please elaborate.

I have an Ansible playbook to deploy a Docker container for Plex Meta Manager which includes all my configuration YAML files (i.e. `config.yml`, and other metadata and collection YAML files). When I want to make a change to my configuration, I change the source files, and re-deploy with Ansible. This will rewrite whatever it's on my server.

You can see how this is problematic for Trakt.tv authentication, which is written to the `config.yml` by PMM. Every time I re-deploy my Ansible playbooks, the `authentication:` section for Trakt.tv will be gone, losing the authentication and refresh tokens, thus losing access to my Trakt.tv data.

### Describe the solution you'd like

Split the Trakt.tv static configuration from dynamic configuration into different files:

`config.yml`

```yml
trakt:
  client_id:
  client_secret:
  pin:
```

`auth.yml`

```yml
trakt:
  authorization:
    access_token:
    token_type:
    expires_in:
    refresh_token:
    scope:
    created_at:
```

This way, I could keep overwriting `config.yml` without worrying that the Trakt.tv authentication details are going to be overwritten and lost.

**Comment by chazlarson** [comment]
Couldn't the Ansible script merge the two files rather than stomping the existing one?  That would solve the problem until such time as this is implemented, if it is.

**Comment by rfgamaral** [comment]
@chazlarson I'm far from an Ansible expert, but I don't think there's a built-in way to do that, not without a complex workaround with multiple steps, I'm afraid. But, again, I'm no expert, maybe there's something I don't know about.

**Comment by chazlarson** [comment]
Maybe useful:
https://www.ansiblepilot.com/articles/concatenate-multiple-files-in-a-specific-order-ansible-module-template-and-yaml/

Doesn't seem really complex on the face of it.  But just a suggestion for a workaround.  I believe the author has said previously he's not interested in putting the auth-related stuff in a separate config.  I can't speak for him, of course.

**Comment by rfgamaral** [comment]
The article demonstrates how to **concatenate** two files on the host, without caring for the content itself. There's added complexity to achieve a working `config.yml` where you want to **merge** (not concatenate) partial content from **a file on the host** and **another file on the remote server**, with care to not overwrite unwanted keys.

I'm not saying it's not possible, I'm pretty sure it is, but there are many steps involved to achieve a good ending result, and I would rather avoid that complexity, if possible. Unless Ansible has a way to smartly merge YAML files, which I don't think it does.

>  I believe the author has said previously he's not interested in putting the auth-related stuff in a separate config.

Oh, that sucks... I'm wondering what's the reasoning behind it, and if in light of the use cased I described in the OP, they would be open to reconsider 🤔

**Comment by chazlarson** [comment]
I was thinking that you'd have the config.yml file with no trakt stuff in it and your auth.yml with only the trakt stuff and you could just concat auth.yml onto the end of config.yml.

If you want generically merge them that could be more complicated.  The Saltbox project does this to bring new settings into the active settings YML file, but it looks like it uses a python script as part of the role:
<https://github.com/saltyorg/Saltbox/tree/master/roles/settings>

**Comment by rfgamaral** [comment]
> I was thinking that you'd have the config.yml file with no trakt stuff in it and your auth.yml with only the trakt stuff and you could just concat auth.yml onto the end of config.yml.

But there's no `auth.yml`, that's pretty much my suggestion on how to improve this issue for automated setups such as the one I have. The `authentication:` bits for Trakt are dynamically generated by PMM, and added to the `config.yml` when running the script.

I could find a way to work around this, but like I said, it would involve a couple of Ansible steps:

- Read the remote `config.yml`
- Extract the Trakt `authentication:` bits
- Merge those bits with my local `config.yml`

It doesn't look that hard or complex, but we can't forget that we are talking about an YAML file here, not a simple text file.

---

Either way, it's not that I can't do it, don't know how to do it, or don't want to do it. I could, but it would definitely be simpler if static configuration and dynamic configuration were separated in their own files, and from an architecture point of view, that's what makes sense to me.

> The Saltbox project does this to bring new settings into the active settings YML file, **but it looks like it uses a python script as part of the role**

You see how this is starting to get "complex" for a thing that could be so painless just by having static and dynamic configuration in different files.

I know you're just trying to help with workarounds, and like I said above, I know it's possible, and I could find a way to do it. But I honestly don't think that's the right approach, and that's why I created this issue. I understand if the author doesn't want to do it, although I don't know their reasons, but I'm hoping they find this a very valid use case.

**Comment by chazlarson** [comment]
Ah, so your use case involves regenerating the auth stuff with every deploy for some reason; or every remote deploy has its own trakt auth that you want to preserve?

I was assuming you had authed against trakt already and just wanted to use that existing authentication in future deploys, not regenerate the auth stuff on every deploy.  That's on me for assuming things not in evidence.

**Comment by chazlarson** [comment]
But, like you said, I'm just trying to offer workarounds for something that is unlikely to be implemented.

The use cases suggested when this has been discussed previously have basically been this very same use case, so it's unlikely that plans will change.  I'll bow out, though.

**Comment by rfgamaral** [comment]
> Ah, so your use case involves regenerating the auth stuff with every deploy for some reason

Not at all, apologies for not being clear.

> I was assuming you had authed against trakt already and just wanted to use that existing authentication in future deploys

That's exactly what I want.

**Comment by rfgamaral** [comment]
> The use cases suggested when this has been discussed previously have basically been this very same use case

But do you know the reasoning behind not doing it?

**Comment by meisnate12** [comment]
once you auth trakt the info should remain the same for some time im just not understanding why you cant deploy that info in the same way you do you tmdb key chaz has a script you can run to just auth trakt and get the values you need

**Comment by chazlarson** [comment]
Then I don't see why the simple concat doesn't work.  You create the config.yml without any trakt stuff in it and the traktstuff.yml that contains only the trakt block and then concat them to config.yml in the ansible role.  If traktstuff.yml doesn't exist the role fails with an error, or not depending on what you want.

**Comment by chazlarson** [comment]
My personal opinion is that implementing this would add more complexity to either everyone's setup [now everyone has to have at least two config files. or maybe not] and the code [since it now has to keep track of what config bits go in which file when it writes them out or whether or not there are one or two to read] for something that benefits only a small portion of the user base.

**Comment by rfgamaral** [comment]
> once you auth trakt the info should remain the same for some time im just not understanding why you cant deploy that info

Ansible works by taking local files (from the machine you're running `ansible`), and applying changes to a **remote machine** through SSH. I don't have the Trakt authentication details locally, they are on the remote server, so I can't just "deploy that info" without reading it first, and finding a way to merge that into my local `config.yml`.

> in the same way you do you tmdb key

The TMDb key is static, it never changes unless I explicitly regenerate a new key manually. Trakt authentication and refresh tokens are dynamic. They are not the same thing.

> Then I don't see why the simple concat doesn't work. You create the config.yml without any trakt stuff in it and the traktstuff.yml that contains only the trakt block and then concat them to config.yml in the ansible role

There's no such thing as `traktstuff.yml`, there's only `config.yml`, with everything in there. It's not possible to concat 2 files, when one of them does not exist. Let me try to explain the workflow...

I have this local `config.yml.j2` (`*.j2` is a jinja2 template file), with something like this:

```yml
~libraries:
  (...)

settings:
  (...)

trakt:
  client_id: MY_TRAKT_CLIENT_ID
  client_secret: MY_TRAKT_SECRET_IT
  pin: MY_TRAKT_PIN
```

When I run my Ansible playbook, this file contents be copied to the remote server, and **fully** replace the `config.yml` in the remote server (where I have PMM running, not the same machine where I'm running the Ansible script).

I then run PMM on the remote server (by logging into the machine with SSH), and it will authenticate against Trakt with the provided client ID and secret. Once that process is done, the `config.yml` in the remote server will be changed to this:

```yml
libraries:
  (...)

settings:
  (...)

trakt:
  client_id: MY_TRAKT_CLIENT_ID
  client_secret: MY_TRAKT_SECRET_IT
  pin:
  authorization:
    access_token: TRAKT_ACCESS_TOKEN
    refresh_token: TRAKT_REFRESH_TOKEN
```

(I have omitted some options for brevity)

Now, let's say, after a few months, a new PMM version comes out with some shinny new settings that I need to tweak. I open my local `config.yml` and change those settings (i.e. the `settings:` section in the file), and then I run the Ansible playbook again. But my local file doesn't have the `authorization:` Trakt section, and once I run the playbook, the remote `config.yml` will be overwritten with my local version, and both the access and refresh token lost. PMM will no longer be able to communicate with Trakt because the authentication information is missing.

I hope that you can see that there's no way to simply concat 2 files. First, I don't have 2 files locally, part of the information is on the remote server. Which means that I would have to first download the remote `config.yml` to my local machine. Then I would need to parse that file, or somehow extract the `authentication:` Trakt section, then append that to my local `config.yml`, and finally push that new file to the server. This is possible through Ansible, I'm sure of it, and that was never the question. But it's very cumbersome to do, and it's very brittle. It's brittle because there's no safe built-in mechanism to do this, it would have to be done with RegExp and whatnot.

> My personal opinion is that implementing this would add more complexity to either everyone's setup [now everyone has to have at least two config files. or maybe not]

Not really. You'd still only have a single config file: `config.yml`. `auth.yml` (or some other name), it would never be a user-configurable file, it would only contain dynamic authentication information that shouldn't be changed by users.

> and the code [since it now has to keep track of what config bits go in which file when it writes them out or whether or not there are one or two to read] for something that benefits only a small portion of the user base.

I honestly don't see how this could be a problem, in the grand scheme of things, this is a minor change. It would not only benefit some users, it's also a best practice to keep **static** and **dynamic** data such as authentication information (that keeps on changing when tokens expire) separate.

**Comment by chazlarson** [comment]
> There's no such thing as traktstuff.yml 

Right, because that's a file you would create to use with this "concat a couple files" ansible workflow.  

Just like PMM doesn't have a `config.yml.j2`; that's a thing you created to use with ansible.

If you don't like the concat thing, you can use the templating system you're already using.

I imagine you are inserting other things into `config.yml.j2` like a plex server URL and token, a TMDB API key, Radarr and Sonarr URLs and API keys, etc.

The trakt details are just another instance of this same thing.

You do the trakt auth one time, and then you store the various values however you are currently keeping track of the rest of the values you're sticking into the template and insert them into `config.yml.j2` the same way you are inserting everything else in there.

It happens that PMM can do that authentication for you if it's not already done, and you provide:
```
  client_id:
  client_secret:
  pin:
```
Doing that every time, of even more than once, is not a requirement, however.  

If you do this authentication today, then capture these values from a config file that contains the filled out trakt block:
```
    access_token:
    token_type:
    expires_in:
    refresh_token:
    scope:
    created_at:
 ```
you can insert them into your `config.yml.j2` like the plex token presumably is.  They don't have to be recreated every time.  You can create them once and then use them later.  I've got a script you can use to generate that block totally independent of PMM if it would be useful, but really all you have to do is go open your current working `config.yml` on the server right now and grab that block of data.

**Comment by rfgamaral** [comment]
That will only work if the `refresh_token` never expires on its own. Trakt documentation does not specify this, but I'm a bit skeptical about it. I think I'll reach out Trakt devs to try and clarify this for me.

However, from a security perspective, that doesn't change the fact that it's not a very good practice to mix authentication data like this in a configuration file. The data should be kept separately, and auth data should have has little permissions as it can.

**Comment by chazlarson** [comment]
I've been using the same block of trakt auth data on three different installs of PMM for at least six months, so it lasts at least that long.

**Comment by rfgamaral** [comment]
The way I'm reading that, you've setup Trakt on one install, and then copied the Trakt authentication details to another 2 installs, correct? If that's what you did, it's strange that that works for that many months.

The access token expires after 3 months (this is on Trakt's documentation), and when that happens, the refresh token is then used to regenerate a new access token. Once the refresh token is used on one of the installs it should expire, and not be usable in the other two installs.

If the refresh token remains usable after being used once, that looks like a security issue on Trakt's side.

**Comment by meisnate12** [comment]
> When I run my Ansible playbook, this file contents be copied to the remote server, and **fully** replace the `config.yml` in the remote server (where I have PMM running, not the same machine where I'm running the Ansible script).
> 
> I then run PMM on the remote server (by logging into the machine with SSH), and it will authenticate against Trakt with the provided client ID and secret. Once that process is done, the `config.yml` in the remote server will be changed to this:

I would recommend authenticating locally and saving that trakt info and push it just like you do for tmdb. Like chaz described above.

> The access token expires after 3 months (this is on Trakt's documentation), and when that happens, the refresh token is then used to regenerate a new access token. Once the refresh token is used on one of the installs it should expire, and not be usable in the other two installs.

In my experience working with the trakt api is that their info on the documentation and what actually happens sometimes differ. I've definitely used access tokens longer then every 3 months



Ultimately im working on other features that would make this change meaningless so theres that too.
